scheds/rust/scx_layered/src/main.rs

```diff
@@ -1376,38 +1525,89 @@ impl<'a> Scheduler<'a> {
+            // $i: The identifier of the metric to set
+            // $e: The expression to set the metric to
+            macro_rules! layer_stat_set {
+                ($i: ident, $e:expr) => {{
+                    let l = self
+                        .open_metrics_stats
+                        .$i
+                        .get_or_create(&vec![("layer_name".to_owned(), spec.name.clone())]);
+                    l.set($e);
+                    l
+                }};
+            }
+            let layer_util = layer_stat_set!(layer_util, stats.layer_utils[lidx] * 100.0);
+            let layer_util_frac = layer_stat_set!(
+                layer_util_frac,
+                calc_frac(stats.layer_utils[lidx], stats.total_util)
             );
-            info!(
-                "  {:<width$}  tot={:7} local={:5.2} open_idle={:5.2} preempt={:5.2} affn_viol={:5.2}",
-                "",
-                ltotal,
-                lstat_pct(bpf_intf::layer_stat_idx_LSTAT_LOCAL),
-                lstat_pct(bpf_intf::layer_stat_idx_LSTAT_OPEN_IDLE),
-                lstat_pct(bpf_intf::layer_stat_idx_LSTAT_PREEMPT),
-                lstat_pct(bpf_intf::layer_stat_idx_LSTAT_AFFN_VIOL),
-                width = header_width,
+            let layer_load = layer_stat_set!(layer_load, stats.layer_loads[lidx]);
+            let layer_load_frac = layer_stat_set!(
+                layer_load_frac,
+                calc_frac(stats.layer_loads[lidx], stats.total_load)
             );
-            info!(
-                "  {:<width$}  cpus={:3} [{:3},{:3}] {}",
-                "",
-                layer.nr_cpus,
-                self.nr_layer_cpus_min_max[lidx].0,
-                self.nr_layer_cpus_min_max[lidx].1,
-                format_bitvec(&layer.cpus),
-                width = header_width
+            let layer_tasks = layer_stat_set!(layer_tasks, stats.nr_layer_tasks[lidx] as i64);
+            let layer_total = layer_stat_set!(layer_total, ltotal as i64);
+            let layer_local =
+                layer_stat_set!(layer_local, lstat_pct(bpf_intf::layer_stat_idx_LSTAT_LOCAL));
+            let layer_open_idle = layer_stat_set!(
+                layer_open_idle,
+                lstat_pct(bpf_intf::layer_stat_idx_LSTAT_OPEN_IDLE)
+            );
+            let layer_preempt = layer_stat_set!(
+                layer_preempt,
+                lstat_pct(bpf_intf::layer_stat_idx_LSTAT_PREEMPT)
+            );
+            let layer_affn_viol = layer_stat_set!(
+                layer_affn_viol,
+                lstat_pct(bpf_intf::layer_stat_idx_LSTAT_AFFN_VIOL)
             );
+            let layer_cur_nr_cpus = layer_stat_set!(layer_cur_nr_cpus, layer.nr_cpus as i64);
+            let layer_min_nr_cpus =
+                layer_stat_set!(layer_min_nr_cpus, self.nr_layer_cpus_min_max[lidx].0 as i64);
+            let layer_max_nr_cpus =
+                layer_stat_set!(layer_max_nr_cpus, self.nr_layer_cpus_min_max[lidx].1 as i64);
```

Comment:
Here too, if shorter names can lead to fewer split statements, it'd be easier on the eyes.

Reply:
Can't the macro just define the left hand variable automatically?

Reply:
Yes, I can use shorter names. I don't think there's any way to define the LHS in a macro. Rust's macro system is referred to as ["hygenic"](https://danielkeep.github.io/tlborm/book/mbe-min-hygiene.html) - this means that variables defined inside of a macro have a different context than those outside the macro. So `let foo = 3` inside a macro cannot be referenced as `foo` in the calling context.

This diverges from C's macro system where you can introduce variables into the calling context. Rust's rules are a bit of a pain in the ass for cases like this, but also avoid a bunch of headaches